src/server.rs

```diff
@@ -553,14 +575,35 @@ impl Server {
         Ok(())
     }
 
+    pub async fn checkin_cleanup(&mut self) -> Result<(), Error> {
+        if self.in_transaction() {
+            self.query("ROLLBACK").await?;
+        }
+
+        if self.needs_cleanup {
+            self.query("DISCARD ALL").await?;
+            self.needs_cleanup = false;
+        }
+
+        self.set_name("pgcat").await?;
+
+        return Ok(());
+    }
+
     /// A shorthand for `SET application_name = $1`.
     #[allow(dead_code)]
     pub async fn set_name(&mut self, name: &str) -> Result<(), Error> {
         if self.application_name != name {
             self.application_name = name.to_string();
-            Ok(self
+            // We don't want `SET application_name` to mark the server connection
+            // as needing cleanup
+            let needs_cleanup_before = self.needs_cleanup;
+
+            let result = Ok(self
                 .query(&format!("SET application_name = '{}'", name))
-                .await?)
+                .await?);
+            self.needs_cleanup = needs_cleanup_before;
+            return result;
```

Comment:
I added a test to verify that we are not sending `DISCARD ALL` after each checkin but it failed because of `set_name` calls. They send `server.query("SET application_name")` call pretty much before each checkin which marks the connection as needing `DISCARD ALL`. This workaround avoids this.